lib/src/rules/always_specify_types.dart

```diff
@@ -164,4 +165,33 @@ class _Visitor extends SimpleAstVisitor<void> {
       rule.reportLintForToken(list.keyword);
     }
   }
+
+  @override
+  void visitMethodInvocation(MethodInvocation node) {
+    if (node.typeArguments == null) {
+      final element = node.methodName.staticElement;
+      if (element is FunctionTypedElement &&
+          element.typeParameters.isNotEmpty &&
+          !_isAllowed(element) &&
+          !element.metadata.any((a) => _isOptionalTypeArgs(a.element))) {
+        rule.reportLint(node.methodName);
+      }
+    }
+  }
+
+  bool _isAllowed(FunctionTypedElement element) {
+    final returnType = element.returnType;
+
+    // Special case for methods with return type and parameter types identical
+    // and equal to the type parameter on function
+    // eg. T m<T>(T p1, T p2);
+    if (returnType is TypeParameterType) {
+      if (returnType.element.enclosingElement == element &&
+          element.parameters.every((p) => returnType == p.type)) {
```

Comment:
I think you're trying to express conditions under which type inference can infer `T`.

If that's true, it ought to be sufficient to have at least one parameter of that type rather than requiring that all parameters are of that type, such as in `T m<T>(T p1, int p2)`. More generally, I think we only need at least one parameter whose type uses that type, such as in `T m<T>(List<T> p1)`.

Reply:
Here I've just implemented what @Hixie suggested in https://github.com/dart-lang/linter/issues/1138#issuecomment-417814244

> The following are cases where the type is used for both the arguments and the return value and it's fine whatever the values are, except that in the case of min and max I'd rather we banned num as an option

So @Hixie is the best person to answer.